examples/greeting-app/src/caching-service.ts

```diff
@@ -30,10 +30,15 @@ export class CachingService {
   ) {
     // Use a view so that we can listen on `refresh` events, which are emitted
     // when the configuration binding is updated in the context.
-    optionsView.on('refresh', async () => {
-      debug('Restarting the service as configuration changes...');
-      await this.stop();
-      await this.start();
+    optionsView.on('refresh', () => {
+      (async () => {
```

Comment:
Hmm, this does not look very pretty, I'd like to take a closer look on what other alternatives we have.

Reply:
Does it imply that event listeners cannot be async functions?

Reply:
@raymondfeng No, it should allow for async listeners. If the framework _did_ accept an async function here, the linter would see the expected type of the callback function was async/promise-returning, and it would not complain here.

The danger with using an async function when it's not expected, is that any rejection will go unhandled, causing the node process to abort. (Unless `onunhandledrejection` has been used.)

This linting helps us to avoid accidentally introducing that danger.

Reply:
@batjos I can see four alternatives:

1. What I did above, with an [async wrapper](https://twitter.com/joeytwiddle/status/1142286305805533185).

2. Same solution but different syntax. (Sometimes I use this syntax, when it looks better, but in this case it looks worse!)

        debug('Restarting the service as configuration changes...');
        this.stop().then(
            () => this.start()
        ).catch(err => ...);

3. Disable the rule for this function only. Without `onunhandledrejection` that would likely have the same result (crashing the process) but it would be less explicit.

4. Disable this part of the rule globally. This would open our code up to the aforementioned danger of unexpected crashing. (But it would keep the checking for promises in conditions, which is valuable enough for me.)

       '@typescript-eslint/no-misused-promises': [
          'error',
          {checksVoidReturn: false},
        ],

Reply:
Thank you @joeytwiddle for a detailed explanation!

I am proposing the following solution:
- Modify `CacheService` to implement `restart` method - this can be an `async` function using `await` keywords
- Modify the event listener as follows:

    ```ts
    optionsView.on('refresh', () => {
      debug('Restarting the service as configuration changes...');
      this.restart().catch(err => {
        console.error('Cannot restart the caching service.', err);
        process.exit(1);
      });
    });
    ```

In the future, we should probably modify `this.restart` to detect the case when another restart is triggered before the first restart has been finished. That's out of scope of this PR though.

Reply:
Hmm yes, async code introduces an entire class of problems and concerns!

Sometimes I find the easiest way to deal with that is to use a queue as a form of locking. In your example, the two restarts would happen one after the other.

In this case it might look something like:

```ts
  private cycleQueue = new TaskQueue();

  // ...

    optionsView.on('refresh', () => {
      this.cycleQueue.queueTask(async () => {
        debug('Restarting the service as configuration changes...');
        await this.restart();
      }).catch(err => {
        console.error('Cannot restart the caching service.', err);
        process.exit(1);
      });
    });
```

(`TaskQueue.ts` is only 38 lines long. Note that it passes the error back out of the queued task to the original caller.)